smartsim/settings/dispatch.py

```diff
@@ -0,0 +1,169 @@
+# BSD 2-Clause License
+#
+# Copyright (c) 2021-2024, Hewlett Packard Enterprise
+# All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions are met:
+#
+# 1. Redistributions of source code must retain the above copyright notice, this
+#    list of conditions and the following disclaimer.
+#
+# 2. Redistributions in binary form must reproduce the above copyright notice,
+#    this list of conditions and the following disclaimer in the documentation
+#    and/or other materials provided with the distribution.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+# DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+from __future__ import annotations
+
+import subprocess as sp
+import typing as t
+import uuid
+
+from smartsim._core.utils import helpers
+from smartsim.types import LaunchedJobID
+
+if t.TYPE_CHECKING:
+    from typing_extensions import Self
+
+    from smartsim.experiment import Experiment
+    from smartsim.settings.builders import LaunchArgBuilder
+
+_T = t.TypeVar("_T")
+_T_contra = t.TypeVar("_T_contra", contravariant=True)
+
+
+@t.final
+class Dispatcher:
+    """A class capable of deciding which launcher type should be used to launch
+    a given settings builder type.
+    """
+
+    def __init__(
+        self,
+        *,
+        dispatch_registry: (
+            t.Mapping[type[LaunchArgBuilder[t.Any]], type[LauncherLike[t.Any]]] | None
+        ) = None,
+    ) -> None:
+        self._dispatch_registry = (
+            dict(dispatch_registry) if dispatch_registry is not None else {}
+        )
+
+    def copy(self) -> Self:
```

Comment:
How is adding type(self) different than just returning the attribute?

Reply:
In this case, not much: it's largely just stylistic! This way if we decide to change the class we don't also have to remember to also change the name of the constructor here